samza-core/src/main/java/org/apache/samza/util/DiagnosticsUtil.java

```diff
@@ -0,0 +1,131 @@
+import java.util.Optional;
+import org.apache.samza.SamzaException;
+import org.apache.samza.config.Config;
+import org.apache.samza.config.ConfigException;
+import org.apache.samza.config.JobConfig;
+import org.apache.samza.config.MetricsConfig;
+import org.apache.samza.config.SystemConfig;
+import org.apache.samza.config.TaskConfigJava;
+import org.apache.samza.diagnostics.DiagnosticsManager;
+import org.apache.samza.metrics.MetricsRegistryMap;
+import org.apache.samza.metrics.reporter.Metrics;
+import org.apache.samza.metrics.reporter.MetricsHeader;
+import org.apache.samza.metrics.reporter.MetricsSnapshot;
+import org.apache.samza.metrics.reporter.MetricsSnapshotReporter;
+import org.apache.samza.runtime.LocalContainerRunner;
+import org.apache.samza.serializers.MetricsSnapshotSerdeV2;
+import org.apache.samza.system.SystemFactory;
+import org.apache.samza.system.SystemProducer;
+import org.apache.samza.system.SystemStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scala.Option;
+import scala.Tuple2;
+import scala.runtime.AbstractFunction0;
+
+
+public class DiagnosticsUtil {
+  private static final Logger log = LoggerFactory.getLogger(DiagnosticsUtil.class);
+
+  // Write a file in the samza.log.dir named {exec-env-container-id}.metadata that contains
+  // metadata about the container such as containerId, jobName, jobId, hostname, timestamp, version info, and others.
+  public static void writeMetadataFile(String jobName, String jobId, String containerId,
+      Optional<String> execEnvContainerId, Config config) {
+
+    Option<File> metadataFile = JobConfig.getMetadataFile(Option.apply(execEnvContainerId.orElse(null)));
+
+    if (metadataFile.isDefined()) {
+
+      StringBuilder metadata = new StringBuilder("Version: 1");
+      metadata.append(System.lineSeparator());
+      MetricsHeader metricsHeader =
+          new MetricsHeader(jobName, jobId, "samza-container-" + containerId, execEnvContainerId.orElse(""), LocalContainerRunner.class.getName(),
+              Util.getTaskClassVersion(config), Util.getSamzaVersion(), Util.getLocalHost().getHostName(),
+              System.currentTimeMillis(), System.currentTimeMillis());
+
+      MetricsSnapshot metricsSnapshot = new MetricsSnapshot(metricsHeader, new Metrics());
+      metadata.append("MetricsSnapshot: ");
+      metadata.append(new String(new MetricsSnapshotSerdeV2().toBytes(metricsSnapshot)));
+      FileUtil.writeToTextFile(metadataFile.get(), metadata.toString(), false);
+    } else {
+      log.info("Skipping writing metadata file.");
+    }
+  }
+
+
+  /**
+   * Create a pair of DiagnosticsManager and Reporter for the given jobName, jobId, containerId, and execEnvContainerId,
+   * if diagnostics is enabled.
+   * execEnvContainerId is the ID assigned to the container by the cluster manager (e.g., YARN).
+   */
+  public static Optional<Tuple2<DiagnosticsManager, MetricsSnapshotReporter>> buildDiagnosticsManager(String jobName, String jobId,
```

Comment:
Should return a java pair class. Can use from apache commons, or add a new one to util, it's generally useful.